### PR TITLE
refactor(protocol-engine): Add provisional command to calibrate gripper

### PR DESCRIPTION
**Pull request by SyntaxColoring**
# Overview

This gives us a basic HTTP interface (via Protocol Engine) to calibrate the gripper. Closes RLAB-191.

# Changelog

* Add a new command, `calibration/calibrateGripper`, that basically just calls `ot3_calibration.calibrate_gripper()` to start capacitively probing.
* Add the new command to the JSON protocol v7 draft. I don't think this is proper, because the new command isn't a protocol command, but this is consistent with #11599, and we can always remove it later.
* Do some minor incidental refactors.

# Testing

I've tested:

* On an OT-2, this raises `UnsupportedHardwareTypeError`.
* On an OT-3 without a gripper, this currently raises an `UnexpectedProtocolError`. (I didn't see a need to make a public error type for this just yet, though we certainly could.)
* On an OT-3 with a gripper, this makes the gripper move.
    * I'm not sure it's returning the correct thing, though. See `# todo` comments.

# Review requests

See the `# todo` comments.

# Risk assessment

Low. Nothing relies on the new code yet, and this doesn't touch existing production code.

### CODE REVIEW

File: api/src/opentrons/protocol_engine/commands/calibration/calibrate_gripper.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        """Execute a `calibrateGripper` command.
+
+        1. Move from the current location to the calibration area on the deck.
+        2. Using the given probe, sense the gripper's offset from its ideal position.
+        3. Return the gripper's offset from its ideal position.
+        """
+        ot3_hardware_api = ensure_ot3_hardware(self._hardware_api)
+
+        # TODO(mm, 2022-11-21): This will error if the gripper jaws haven't been homed.
+        # Does it make sense to home the jaws automatically?
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/commands/calibration/calibrate_gripper.py]: Ya, I think the hardware controller should home jaws before calibration if they aren't homed.
What do you think @ahiuchingau? I think we can just home jaws, if they aren't already, before calling `grip` in `calibrate_gripper`.

File: api/src/opentrons/protocol_engine/types.py [hunk 2]
```diff
@@ -203,6 +205,15 @@ class ModuleDimensions(BaseModel):
     lidHeight: Optional[float]
 
 
+class Vec3f(BaseModel):
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/types.py:218]: I'm guessing we can remove this and use `DeckPoint` if the mount calibration function returns an absolute deck location?

**Review by sanni-t — COMMENTED** [review]
LGTM! Pending answer about the calibration offset

File: api/src/opentrons/protocol_engine/types.py (hunk 2 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/types.py:218]: We could, but I'd prefer to merge `DeckPoint` with `LabwareOffsetVector` and `InstrumentOffsetVector` even then.

Although they all represent semantically different things, I don't think it's worthwhile for them all to be separate types. Basically for the same reason we don't define a separate `float` type every time we want to use a `float` in a model.

File: api/src/opentrons/protocol_engine/commands/calibration/calibrate_gripper.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        result = await ot3_calibration.calibrate_gripper(
+            ot3_hardware_api, self._convert_to_hw_api_probe(params.probe)
+        )
+
+        # TODO(mm, 2022-11-21): Clarify what calibrate_gripper() returns--absolute
+        # coordinates, or an difference from the ideal? Is it correct to return its raw
+        # result here?
+        return CalibrateGripperResult.construct(
+            probeOffset=Vec3f.construct(x=result.x, y=result.y, z=result.z)
```
Comment (sanni-t) [inline api/src/opentrons/protocol_engine/commands/calibration/calibrate_gripper.py]: [This recent PR](https://github.com/Opentrons/opentrons/pull/11786/files) updated the return value of `calibrate_gripper` to be the offset, i.e., the difference from ideal. So this looks correct.

File: api/src/opentrons/protocol_engine/commands/calibration/calibrate_gripper.py (hunk 1 above)
Comment (sanni-t) [reply api/src/opentrons/protocol_engine/commands/calibration/calibrate_gripper.py]: We'll assume that hardware controller will home all axes before starting calibration. So this is good as is.

File: api/src/opentrons/protocol_engine/commands/calibration/calibrate_gripper.py (hunk 3 above)
Comment (SyntaxColoring) [reply api/src/opentrons/protocol_engine/commands/calibration/calibrate_gripper.py]: Thanks for looking into that!

File: api/src/opentrons/protocol_engine/commands/calibration/calibrate_gripper.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+        # TODO(mm, 2022-11-21): This will error if the gripper jaws haven't been homed.
+        # Does it make sense to home the jaws automatically?
+        result = await ot3_calibration.calibrate_gripper(
+            ot3_hardware_api, self._convert_to_hw_api_probe(params.probe)
+        )
+
+        # TODO(mm, 2022-11-21): Clarify what calibrate_gripper() returns--absolute
+        # coordinates, or an difference from the ideal? Is it correct to return its raw
+        # result here?
```
Comment (SyntaxColoring) [inline api/src/opentrons/protocol_engine/commands/calibration/calibrate_gripper.py]:
```suggestion
```

**Review by sanni-t — APPROVED** [review]
📏📐

**Pull request merged**